### PR TITLE
#1200: share button for extensions

### LINKED ISSUE #1200: Add a "Share" button for extensions
**Context**
- We have an "Export as Blueprint" modal that 

**Wireframe**
- See the "Share" wireframe

**Implementation Sketch**
- Add a Share button
- Extract the blueprint adapter from "Export as Blueprint" 
- Add a modal to input the metadata for the blueprint
- Use the ShareTable from the workshop

**Additional Considerations**
- Need to handle also sharing dependencies

### PR DESCRIPTION
**Pull request by twschiller**
Closes #1200

Considerations:
- Defines a "share" route (made it easier to reload the page during development)
- Defines a HorizontalFormGroup helper component for rendering the components in a standard way. We should start thinking about what the right abstraction is for re-usable form fields that show error state properly
- Passes a share_dependencies flag so that dependencies are automatically shared too https://github.com/pixiebrix/pixiebrix-app/issues/527

Out of scope:
- Implement displaying dependency information

image

### CODE REVIEW

File: src/components/ellipsisMenu/EllipsisMenu.tsx [hunk 1]
```diff
@@ -37,17 +38,20 @@ const EllipsisMenu: React.FunctionComponent<{
...
-            item.action();
-          }}
-          className={item.className}
-        >
-          {item.title}
-        </Dropdown.Item>
-      ))}
+      {items.map(
+        (item, index) =>
+          !item.hide && (
```
Comment (BALEHOK) [inline src/components/ellipsisMenu/EllipsisMenu.tsx:43]: how about `items.filter(item => !item.hide).map...` ?

File: src/options/pages/installed/InstalledPage.tsx [hunk 2]
```diff
@@ -42,15 +42,25 @@ import { RemoveAction } from "@/options/pages/installed/installedPageTypes";
...
-}> = ({ extensions, onRemove }) => {
+}> = ({ extensions, onRemove, push }) => {
   const { flags } = useContext(AuthContext);
 
+  const matchShare = useRouteMatch<{ extensionId: string }>({
+    path: "/extensions/share/:extensionId",
+    strict: true,
+    sensitive: true,
+  });
+
```
Comment (BALEHOK) [inline src/options/pages/installed/InstalledPage.tsx:63]: is it possible to use `Route` instead of manual matching?

File: src/utils/objToYaml.ts [hunk 3]
```diff
@@ -19,7 +19,8 @@ import { sortBy, fromPairs } from "lodash";
 import { removeUndefined } from "@/utils";
 import { dump } from "js-yaml";
 
-export const objToYaml = (obj: Record<string, unknown>) =>
+// eslint-disable-next-line @typescript-eslint/ban-types -- don't need/want index signature
+export const objToYaml = (obj: object) =>
```
Comment (BALEHOK) [inline src/utils/objToYaml.ts:23]: could you elaborate on the reason for this change?

**Review by BALEHOK — CHANGES_REQUESTED** [review]

File: src/options/pages/installed/InstalledPage.tsx (hunk 2 above)
Comment (BALEHOK) [reply src/options/pages/installed/InstalledPage.tsx:63]: like this (from line 109)
```
<Page title="Active Bricks" icon={faCubes}>
  <Route path="/extensions/share/:extensionId" exact>
    <ShareExtensionModal
      extension={toShare}
      onCancel={() => {
        push("/installed");
      }}
    />
  </Route>
...
```

File: src/components/ellipsisMenu/EllipsisMenu.tsx (hunk 1 above)
Comment (twschiller) [reply src/components/ellipsisMenu/EllipsisMenu.tsx:43]: I think that might screw up the `key={index}`? Not sure - thought I needed consistent use of indexes

Comment (twschiller) [reply src/components/ellipsisMenu/EllipsisMenu.tsx:43]: Will change for now and see if we hit problems

**Review by BALEHOK — APPROVED** [review]

**Pull request merged**